### ISSUE TITLE
Detect Moves enhancements

### ISSUE THREAD

**Issue by pnedev** [pnedev/compare-plugin#165]
Continued from https://github.com/pnedev/compare-plugin/issues/113#issuecomment-383124079

**Comment by pnedev** [comment]
Hello Yaron,

Some brief comments.

> The problem is I don't remember the moved-bugs we've found in 1.5.6.6.

I didn't find exactly bugs AFAIR. Perhaps some sub-blocks were not marked as moved.
I wanted to add multiple moved indication just to let the user know that the block appears 'as-is' in several places.

> Certainly IMO. And this is how it's implemented in previous versions.

OK, fair enough. Although in text files with reoccurring patterns (source code `if (x)` lines that often appear) it is very uninformative to mark those as moved in the scope of a large diff block.
This is what makes the _Detect Moves_ implementation so controversial - it depends on the file contents type. The same applies to _Changed detection_ as well.

> Changed must precede/override Moved.

This is a dangerous path. In your example only one or two changes exist between the lines. There might be quite many changes as well. And this might be a preference / settings file which lines are unique but the words in the different lines are common.

> Indeed. Or similar to Multiple-Moved.

OK, got it.

> Yes, it might be too much for some users...

Let's not complicate thing further with another setting. Let's just try and add the `+.` / `-.`  marks. I'd like to have a reminder that a line/block is _moved-multiple/questionably-moved_ and the icons you suggest provide such reminder while the line color still signifies clearly that the line/block is Added/Removed.

--------------------------

The new moved block icons are added in **[this x86](https://ci.appveyor.com/api/buildjobs/d09d360gubi9aqt2/artifacts/ComparePlugin.dll)** binary, thanks for providing those.

Thank you for the feedback Yaron, appreciated as always.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

> I didn't find exactly bugs AFAIR.

That's good. :)

> I wanted to add multiple moved indication just to let the user know that the block appears 'as-is' in several places.

That was a great idea.

> Although in text files...

The context/block-detection *concept* may also strengthen your approach to some degree.
That is: just like the context can SET a Move, it can also indicate that two identical lines are NOT a Move.
But this is too sophisticated and I'm glad you've agreed to set a Move in this case. 

> Changed must precede/override Moved.
...
This is a dangerous path. In your example only one or two changes exist...

You're absolutely right.
I was wrong both practically and also in terms of concept; Move can override Added/Removed, so why not Changed?
I must compliment you again for this analysis. :+1: 

> Let's just try and add the +. / -. marks.

Great.
If I understand you correctly, you mean to implement this regardless of Detect Moves.
[Here](https://github.com/pnedev/compare-plugin/files/1940112/LocalAddedRemoved.zip) are the icons. Whenever you find the time.    

Thanks for updating the Block-Moved icons.
And thanks again for your brilliant work. I do appreciate it.

BR

**Comment by pnedev** [comment]
Hello Yaron,

Thank you for the feedback, much appreciated.
And thanks for the icons, great job as always. :+1: 

> If I understand you correctly, you mean to implement this regardless of Detect Moves.

No, it is linked to move detection and it will complement it. It is kind of 'moved-multiple', right?

BR

**Comment by Yaron10** [comment]
Hello Pavel,

> No, it is linked to move detection and it will complement it. It is kind of 'moved-multiple', right?

It can also be an independent feature.
Both approaches are fine and I wasn't sure what you meant. Thanks for clarifying.

Best of luck with the re-work. I appreciate it.

**Comment by pnedev** [comment]
Hello Yaron,

The smart move detection is ready, here is the **[x86](https://ci.appveyor.com/api/buildjobs/f7qkarqgivmvav7w/artifacts/ComparePlugin.dll)** build.
Here are also several test files (you probably have them but I'm including them for convenience) - 
[move_examples.zip](https://github.com/pnedev/compare-plugin/files/1997438/move_examples.zip).

Thank you once again for the help.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

This is one of your major improvements and, perhaps, the most important of them.
Thank you very much for your really smart work. :+1: :+1: 

Some comments with your permission.

I think that `G` (line 15 in abc.txt, pair3) and `test test test` (line 25 in prefsFF.js, pair 4) should be marked as Local-Added/Removed.
Unlike Move, we shouldn't care about uniqueness here. On the contrary - this is a complementary feature resembling Multiple-Move which means "no uniqueness".
And the rule IMO should be: if the line exists in the other file (*including matches*) and it can't be marked as Moved, mark it Local-Added/Removed.

The theoretical feature "Go to Moved Line/Block in the Other File" can be a good indication to whether or not a line/block should be marked as Moved: it should make sense that the lines/blocks in both files are indeed a pair.
But this is irrelevant in Added/Removed. I just want to know whether "Removed" means "in the entire other file" or "only here".      

**
Shouldn't empty lines be marked Local/Global as well?

**
Allow me to return to your "I am single moved line in completely different block. Should I be marked as moved?" question.

Testing [these](https://github.com/pnedev/compare-plugin/files/1997930/TEST.zip) files, I understand you point and approach better.
Lines 879 and 471 (englishOld.xml and englishNew.xml respectively) are marked as Moved.
They are not a pair: 879  is under "Run a Macro Multiple Times" and 471 is under "ShortcutMapperSubDialg".

How about this?
If a line appears *once* in each file it's a certain Move; we don't have a leeway and it must be marked as Moved even if the context is completely different.
But if a line appears multiple times we're in "question zone" and smart-detection can be applied *even if there's only one occurrence of the lines besides matches*. 
IOW - the rule should be simpler: if a line appears multiple times, it should be marked as Moved only with 
smart/context-detection (as opposed to the current implementation that matches are taken into account).

So in your `I am single moved line in completely different block.` files those lines should be marked as Moved.
But `<Item id="2" name="Cancel"/>` in the language files should be Local-Added/Removed.

You deserve another standing ovation. :)
Thanks again.

BR

**Comment by pnedev** [comment]
Hello Yaron,

Thank you for the quick feedback and the clarifications. :+1: 

> And the rule IMO should be: if the line exists in the other file (including matches) and it can't be marked as Moved, mark it Local-Added/Removed.

What exactly do you mean by _including matches_?
The line `G` from pair3 is marked as moved (in a block context) but still you suggest to mark it as locally removed?! Do you mean that ONLY lines that _single-match_ (do not match in a block context but as single lines) should be excluded from locally added/removed groups?

> Shouldn't empty lines be marked Local/Global as well?

I think they shouldn't. Blank lines are quite common in text files and almost always they are 'cosmetic' - don't have functional significance. Thus, it is completely unnecessary to mark them as local/global - doing so will only mess things up. The only exclusion is when they are part of a block-match. Then they are part of the block and are marked as the block is - either moved or locally added/removed (context match as you called it).

> IOW - the rule should be simpler: if a line appears multiple times, it should be marked as Moved only with
smart/context-detection (as opposed to the current implementation that matches are taken into account).

I didn't get that too, sorry.

P.S. I'm sorry if I sound brief but I'm quite busy these days.

Thank you once again,
BR

**Comment by Yaron10** [comment]
Hello Pavel,

> The line `G` from pair3 is marked as moved...

I was referring to line **15** and forgot to mention line 9 as well.

![default](https://user-images.githubusercontent.com/10229320/40002243-c3712490-5798-11e8-8502-ad0b97bf3320.png)

If a line exists *anywhere* in the other file and you can't mark it Moved *in this file*, it should be Local.
Block/context is irrelevant to Added/Removed.

**

If you agree with my previous comment, the following Global-Added is highly confusing.

![default](https://user-images.githubusercontent.com/10229320/40002968-db30f270-579a-11e8-81a0-0e954d82f95d.png)

Think about the user-guide: you explain to the user that ![default](https://user-images.githubusercontent.com/10229320/40003028-fc0711c8-579a-11e8-936a-34d2b71cac7e.png) means "This line was added in this file and does not exist anywhere in the other file.".
Why complicate it and mark empty lines differently?

**
> Allow me to return to your "I am single moved line in completely different block."...
...

I realized that your approach made a lot of sense (well, not for the first time :) ).
I just think that if a line appears *once* in each file, it should override the different-context and be marked as Moved. 
***
Your time is highly appreciated. Whenever you can.

BR

**Comment by Yaron10** [comment]
Hello again Pavel,

Allow me a few more lines regarding the first two points.

Up until now Added meant: "The line was added in this location {globally - unknown}".
You can think of ![default](https://user-images.githubusercontent.com/10229320/40011413-72ae07ae-57b0-11e8-84f3-ca30ccbb0af1.png) as the new "default" Added: "The line was added in this location {globally - known}".

I'd appreciate it if you could find the time to let me know if my points are clear.
Thank you.

BR

**Comment by pnedev** [comment]
Hello Yaron,

Yes, your points are quite clear now, thanks for the explanation.
However, this is quite new concept and I need to consider it carefully.

To be able to tell if a line does exist _anywhere_ in the other file I need to actually search for it. So I'll have to compare ALL lines with each other. This is completely new functionality and it might slow the compare operation. You'd like to actually combine the _Compare_ and _Find Unique_ commands in one (making _Find Unique_ unnecessary) this is how I get it.

If you compare settings files that approach will be useful but if you compare source files (that have reoccurring patterns) it doesn't really bring much value (a `if`, `while`, etc. lines are globally irrelevant). This is true for all kind of text files that are somewhat context based. And most of the text files are context based.

You are right that now _Added/Removed_ is by default local (and globally unknown) but this is exactly what compare operation does. It finds _local_ changes and context mismatches. It is the compare functionality. The other thing is search functionality. It is better for compare process to concentrate ONLY on the changes and how they relate to each other and 'forget' about the matches.

Even if I implement it then why use the new 'plus-dot' / 'minus-dot' marks? It is the default meaning (local match) and presumably it will occur more often so why not use the default and simpler symbols? And make another "enhanced" symbols that will stress the uniqueness of the line in a file? That will outline (stress) the fact that a line is MORE than just added/removed locally (as are all others).

I'm not saying that all this is bad but it will require some serious thinking and a major rework.

Does it mean that a line/block will be considered _Moved_ ONLY if it matches once? If the line is found N times in file 1 and N times in file 2 (N > 1) shall it be considered _Moved_ as well (as in the current implementation)?

Blank lines appear quite often (at least in the type of files I'm working with) and don't bring any information. They are simply separators for readability (mostly). Comparing them and matching them with each other is not quite useful and informative in this regard so I don't find it justified to mark them as _Moved_ or _locally_ changed unless they are part of a larger context/block (then I include them to the block to keep the block color mark consistent). Imagine a large added block that contains single blank line which is marked as moved because it matches removed blank from the other file.
Having said that I need to mention that the move detection algorithm handles blank lines this way. I haven't yet found a reasonable way to implement _Ignore Blank Lines_ option and the basic compare operation treats blanks the same way as the other lines.

Thank you for commenting, I appreciate it :+1: 

BR

**Comment by pnedev** [comment]
Hello Yaron,

> IOW - the rule should be simpler: if a line appears multiple times, it should be marked as Moved only with
smart/context-detection (as opposed to the current implementation that matches are taken into account).

I thought about that and I see what you mean. That is a clever idea indeed, very good :+1: 
I have implemented it and I have also fixed another issue (I was erroneously treating locally added/removed lines as moved - they didn't take part in the line changes detection process).

**[Here](https://ci.appveyor.com/api/buildjobs/10e2iqh01m42hhar/artifacts/ComparePlugin.dll)** is the latest x86 binary.

Thank you,
BR

**Comment by Yaron10** [comment]
Hello Pavel,

Thank you for the detailed explanation. I appreciate it.

I'm glad we're on the same page now.
This is the concept I had in mind when I first suggested it.
Our discussion was interrupted by the important blank rework. 
***
> I'm not saying that all this is bad but it will require some serious thinking and a major rework.

I'm retracting my request. :)
It's up to the importance you see in it, your energy and enthusiasm.
How about leaving it as an option for a future version? https://github.com/pnedev/compare-plugin/issues/141 and https://github.com/pnedev/compare-plugin/issues/138 are certainly more important for a stable new release.
Anyway, many thanks for considering it. 

So the following lines are for the future. Or purely intellectual. :)

> So I'll have to compare ALL lines with each other. This is completely new functionality and it might slow the compare operation.

IMO it's worth a setting.

> You'd like to actually combine the Compare and Find Unique commands in one (making Find Unique unnecessary) this is how I get it.

1. "Find Unique" displays unique lines much more clearly.
2. "Find Unique" covers *Changed lines* as well.

> Even if I implement it then why use the new 'plus-dot' / 'minus-dot' marks?

It's very hard to play with 14x14 images. Reducing the plus and adding a dot was easier.
But obviously we can discuss that "if and when".

IMO excluding empty lines from Local/Global would be confusing.
Practically in most cases they would be Local.

**

Just to make sure: if you decide not to implement "Full Local/Global", we return to the old plus/minus icon and the new ones should be removed. Is that correct?
***

> Does it mean that a line/block will be considered Moved ONLY if it matches once? If the line is found N times in file 1 and N times in file 2 (N > 1) shall it be considered Moved as well (as in the current implementation)?

I assume this is not related to "Local/Global Added/Removed" but rather to core Move.
The current implementation is very good and "If the line is found N times in file 1 and N times in file 2 (N > 1)" it should be marked as Moved only with smart/context detection.

I suggested one change though: currently if a line is found twice in both files, one is a match and the other in different locations - it's marked as Moved regardless of the context.
I think it should depend depend on the the context (this is your original approach).

But if a line is found *once* in both files, it should be marked as Moved regardless of the context.
In the rare case that both file contain a single empty line in different locations - it should be marked as Moved IMO. It might have significance.
In the more common cases (multiple empty lines), smart/context detection is required (as with regular lines).
***
As I was writing those lines my Email-Notifier informed me of your new post. Thank you.
I'll write again later.

BR

**Comment by Yaron10** [comment]
Hello again Pavel,

> IOW - the rule should be simpler: if a line appears multiple times, it should be marked as Moved only with
smart/context-detection (as opposed to the current implementation that matches are taken into account).
...
I thought about that and I see what you mean. That is a clever idea indeed, very good

Thank you. You deserve the credit as it was your original approach.

But do we mean the same thing? :)
Could you please compare the [language files](https://github.com/pnedev/compare-plugin/files/1997930/TEST.zip)? 
Lines 879 and 471 (englishOld.xml and englishNew.xml respectively) shouldn't be marked as Moved.

> I have also fixed another issue (I was erroneously treating locally added/removed lines as moved - they didn't take part in the line changes detection process).

Thank you for that too. I don't understand the problem and the fix though.
Regardless, do you see any point in keeping Local/Global if it's not fully implemented?
IMO it should be removed.

I appreciate your work. Thanks again.
BR

**Comment by Yaron10** [comment]
Hello Pavel,

I am sorry about the misunderstanding.
And thanks again for the detailed explanation.

> It looks like a good complementary feature but I have to consider it well and all the implications it might have over the basic functionality. I personally have never needed such feature.

So possibly in the future. :)
Thanks.

>  In this regard the locally added/removed means "Hey, that diff is not exactly moved but it appears in the other file's diffs. IOW - it is not just added/removed in this file - it is also removed/added in the other file".

Well, Poirot had to use the gray cells. I hope I got it correctly.

This is essentially multiple-move or question-move, isn't it?
Marking multiple-move can certainly be useful; but we (that is: you :) ) have tried a multiple-move icon, block-mode and line-mode and decided not to use any. 
And I'm afraid the current implementation is not better.

1) In `+` your reference is local only (globally unknown). I find it confusing that in `+.` (something like `+!` actually might be more suitable in your concept) the reference is to a different location.  
2) Your definition of `+.` is: "This line has a paired `-.` in the other file.".
Besides being too sophisticated IMO, it *essentially* means: "This is a question-move.". Which, in turn, should mean that `+` is certainly NOT a question-move. And this is not true.
3) More simply: implementing multiple-move should be clearly Move (blue highlight and a similar icon) and not mixed with added/removed.

This is my opinion. If you accept it, I apologize for your extra work (which is truly appreciated anyway).
***
> In the current implementation it ALWAYS depends on the context.

We're not entirely on the same page. :)
`I am single moved line in completely different block. Should I be marked as moved?` is marked as Moved.

> Compare pair4.
Shall G-E blocks be marked all as moved because they match equal times (as contexts)?

This is interesting: you get different results depending on which file is Old and which is New.

If `1.txt` is New, `G-E` should not be marked as Moved.
`G-E` (in two lines) should be treated as `GE` (in one line). A block is basically a long line and the context for `GE` does not indicate a Move.
IOW: if you select the two lines `G` and `E` and open the incremental-bar you get *two* matches.
You can mark a Move based on context only if there's *one* match (including matches or not is the question, but certainly not two Moves).

If `2.txt` is New, `G-E` and `A-B` are marked as Moved.
In v1.5.6 they are also marked as Moved. Reason/Rationale: the other `G-E` and `A-B` are matches.
My suggestion was to change that and not mark as Moved in such cases.
Rationale: Moved-mark/icon is helpful in some cases but can be confusing in others (as you have very well explained to me :) ). 
Line-Mode / Block Mode is removed. We don't know the contents of the files and what would actually be better. And therefore - whenever in doubt and possible (i.e. unlike `I am single moved line`) do not mark as Moved. 

**

> "found once in both files" means exactly that for you which is "once in the WHOLE files";

I actually meant "found once in each file". :)

> Compare pair5.
Lines G are marked as locally added/removed because they match several times AMONG THE EXISTING DIFFS.

Unless we find a better way to mark question-move, I think it should be added/removed.

> Compare pair6.
Lines G are marked as moved because they match once IN THE EXISTING DIFFS and regardless of the fact that unchanged line G exists in both files (line 1).

This is the "including matches or not" question. :) 
(pair4, `2.txt` is New).

The issue is rather complex. I hope my calculations and reasoning are correct. 
***
> #141 and #138 are certainly more important for a stable new release.
...
It is a bit off-topic but those two issues might remain unfixed for the new release. There is no way to circumvent them, at least not a straightforward one.

Those issues are major and even critical IMO.
I've found some cases we haven't discussed yet (e.g. NPP Search functionalities don't work properly in large files).
I didn't want to bother you with that in the middle of the Move rework.
With your permission I'll elaborate on Friday (I won't be by my PC tomorrow).

Thank you very much. Truly and highly appreciated.

BR

**Comment by Yaron10** [comment]
I've edited the following lines for clarity.

> IOW: if you select the two lines `G` and `E` and open the incremental-bar you get *two* matches.
You can mark a Move based on context only if there's *one* match (including matches or not is the question, but certainly not two Moves).

to

> IOW: if you select the two lines G and E and open the incremental-bar you get two search-matches.
You can mark a Move based on context only if there's one search-match (including CP-matches or not is the question, but certainly not two Moves).


Thanks again.
BR

**Comment by pnedev** [comment]
Hello Yaron,

Thank you for sharing your perspective.
I understand you reasoning better (at least on most points :))

You got it right - currently it is the _moved-multiple_ concept.

I need to be brief.

--------------------------------

Because of the complexity of the issue and the diverse assumptions and views I think it's best to go back to the very basics:

1) After compare we have a list of diff sections - added and removed.
2) If the user desires move detection (the option is on) then check all added diffs for removed counterparts in the other file.
3) If such match is found and it is unique (within the diffs of course because they are the changes between the files) mark the pair as moved.

I think it is rather straightforward.
There are several details however.

- This is regarding point 2 above.
Should the diffs be matched as a whole?

1) That will be the most certain - one added block matches one removed block entirely.
2) The other approach would be to match parts of the diffs (as it is now) but then we might end up with an added line that exist in several removed blocks. Which one is the moved?
Well we do our best and look at the surrounding DIFF lines. If we detect that the added line and several of its surrounding lines match the same removed lines in removed diff and that match is unique then mark this as a moved sub-block.

- This is regarding point 3 above.
How do we define uniqueness?
If there is only one match then there's no question - it is unique.
If however the same added section is paired several times (that is, all added sections X have their removed counterparts) then we can say for sure that all are moved.

Do you find any ambiguity or inconsistency in such approach?
The move detection simply extends the compare process by diffs / sub-diffs matching and states the facts. It doesn't interpret anything.

So I think it's best to return to the original move detection with the addition of the several equal move matches.

--------------------------------

> So possibly in the future.

Now your suggestion fits perfectly as another extension - lines uniqueness matching based on another user option. And now I understand why it is not actually connected to move detection as you initially stated. :+1: 

--------------------------------

> Those issues are major and even critical IMO.
I've found some cases we haven't discussed yet (e.g. NPP Search functionalities don't work properly in large files).

That's interesting. I would like to here more about it, thank you very much.

I will be travelling for a week, I'll write as soon as I can.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

> If however the same added section is paired several times (that is, all added sections X have their removed counterparts) then we can say for sure that all are moved.

I hope I understand you correctly.

Re-implementing Move-Multiple can be helpful in many cases.
(I thought that my Local/Global suggestion would *indirectly* help in Move-Multiple too and make the explicit implementation unnecessary; but now we're back in square one :) ). 
But wouldn't it be better to have it as an option (something like `Use extended Move marking`)? 
In many other cases - too many Move marks would be distracting and confusing.

What do you think?
***
> If such match is found and it is unique (within the diffs of course because they are the changes between the files) mark the pair as moved.

If we term Move-Multiple "Extended Move Marking", the other method should be "Minimal Move Marking"; - whenever in doubt, do not mark as Moved.

I think that for good practical Move marking in real life (i.e. not too many confusing marks), matches should not be excluded when checking for uniqueness.

And the rule for "Minimal Move Marking" should be: if you select the Moved line/block and open the Incremental-Bar, you should get *one* search-match in each file.
(If `Test` is marked as Moved and you also have `//Test`, you'll get two search-matchs; but I mean one search-match for the Moved line/block as is).

Even if you don't agree, `G-E` in pair4 should not be marked as Moved.  

Thank you. I appreciate your work and patience.

Have a nice weekend and enjoy your traveling.
BR

**Comment by pnedev** [comment]
Hello Yaron,

Yes, here is shortly what I'm planning to do:

1. Remove multiple move concept altogether.
2. Treat and mark all same matched diff lines (if the match count between the files is equal and the context is the same) as moved. By 'context' here I mean if the other diff lines, surrounding the match line also match. Pair4 is a good illustration. This means that if you search for the whole block (match line + its context) you should find it equal times in both files.
3. I'll implement (+user option) your concept of uniqueness -> locally added/removed lines that will complement the basic move algorithm. So all unmoved diff lines will be checked for local or global addition/deletion.

This way we'll have the old move approach (plus the equal number matching) and optionally (user decision) the local/global detection for the other diffs as extension.

How does it sound?

Thanks.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

Thanks for replying. I appreciate it.

> Remove multiple move concept altogether.

OK.

> Treat and mark all same matched diff lines (if the match count between the files is equal and the context is the same) as moved. By 'context' here I mean if the other diff lines, surrounding the match line also match. Pair4 is a good illustration. This means that if you search for the whole block (match line + its context) you should find it equal times in both files.

I'm not sure I understand.

Should `G-E` in pair4 be marked as Moved?
IMO they *certainly should not*. This is fundamentally contrary to the concept.
Single lines (i.e. not blocks) are marked as Moved only if they appear *once in each file* (only diffs or entire files is another question).
Why should blocks be different? The rule should be the same: a block is marked as Moved only if it appears *once in each file*.

Am I missing anything? 

> I'll implement (+user option) your concept of uniqueness -> locally added/removed lines that will complement the basic move algorithm. So all unmoved diff lines will be checked for local or global addition/deletion.

:+1: 
That's absolutely great! Thank you very much.
Do you mean to mark Changed lines as global/local as well (i.e. "Changed here but exists as is in some other location in the other file."? - That would be excellent; we would need two more icons, wouldn't we?

`Find Unique` would still be a great feature as it's much clearer for that purpose.

I highly appreciate your time and work.
BR

**Comment by pnedev** [comment]
Hello Yaron,

> Should G-E in pair4 be marked as Moved? ......

Yes. At least IMO.
Let's compare again pair4 with the latest CP, file 1 is new.
Looking only at file 1 what do we see? G-E block at line 6 is moved and G-E block at line 13 is also moved. Are they not? They are not simply added - they exist exactly as they are in file 2, just in different positions. Looking only at file 1 I instantly know that the change is only reordering of some lines - nothing is actually added or removed - the difference between the files is that some lines have changed their position. That's it. We can tell that for sure. Why wouldn't we?

> Do you mean to mark Changed lines as global/local as well ....

Hm, I haven't thought about that but that's true. Then yes - we'll need ... one more icon?

Thank you.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

> Hm, I haven't thought about that but that's true.

:+1: 
You wrote the Local/Global implementation should be hard and time consuming.
Special thanks. I'm sure many other users will appreciate it too. 

> Then yes - we'll need ... [one more icon](https://github.com/pnedev/compare-plugin/files/2058333/ChangedLocal.zip).

Sure, my mistake.
I don't quite like the position/proportion of the slash with the smaller lines (it took me some time to get it right with the older one).
If you think it's too bad, I'll try to improve it.
***
> Let's compare again pair4 with the latest CP, file 1 is new.
...

Your argument makes sense and it's certainly a legit approach.

But... This is *Move-Multiple*.
If you don't implement Move-Multiple, there shouldn't be a difference between `GE` (one line) and `G-E` (two lines).
Any Moved-Line/Block selected by the user should have *one search-match* in each file (entire file or only diffs is another question).

The command `Go to Moved Block in Other File` is also a good indication: you can not know where to go in this case.

Another point: in our discussion we've used "context" quite often; line 10 is equal in both files and this enables us to determine that line 9 above 10 is the correct Moved.
But it's actually much simpler: checking the surrounding diffs, we're trying to find a "new" string (block) which exists *only once in each file*. 

And now practically: currently you get many *unrelated*
![default](https://user-images.githubusercontent.com/10229320/40786687-5162b12a-64f4-11e8-96d8-44e59a64f3c6.png)
when comparing source files; this is extremely confusing and distracting. 

I hope my arguments are clear. :)

Thanks again.
BR

**Comment by pnedev** [comment]
Hello Yaron,

Thank you for the icon, I'll write back when I test it.

> But... This is Move-Multiple.

It is not, at least the way I get it. Move-Multiple is when the match counts in the files are different. Then we cannot know where the line is moved.

> Any Moved-Line/Block selected by the user should have one search-match in each file

My correction would be: _Any Moved-Line/Block selected by the user should have EQUAL search-match in each file_.

> ... you can not know where to go in this case.

It is irrelevant - as it is the same it is moved either here or there but it doesn't matter because it is the same.

> when comparing source files; this is extremely confusing and distracting.

Now that is an argument I can agree with :).
But it affects/relates only to source files isn't it? Although it is not quite useful and informative if you get a common line matching 10 times and you mark it as moved - it is obviously commonly appearing and not unique to be a good candidate for move mark.

Thank you.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

> Now that is an argument I can agree with :).

I thought this would push the right button. :)

> But it affects/relates only to source files isn't it?

Not necessarily. 

> Although it is not quite useful and informative if you get a common line matching 10 times and you mark it as moved - it is obviously commonly appearing and not unique to be a good candidate for move mark.

I'm not sure what your conclusion is.

If you mean keeping the current behavior and mark `G-E` as Moved, I'll relate to the other points and try my best to convince you. :)

If you mean changing the current behavior and mark `G-E` as Added/Removed (Local), what do you think about "once in the entire file or only in diffs?".

Thank you.
BR

**Comment by pnedev** [comment]
Hello Yaron,

I'll mark it moved if it matches only once. You are good at diplomacy ;)
I'm thinking of "once only in diffs" approach. That's about the unique moved right?

BR

**Comment by Yaron10** [comment]
Hello Pavel,

> I'll mark it moved if it matches only once.

:+1: 
Thank you. I appreciate it.

> You are good at diplomacy ;)

Far from it.
My intuition is good though and you are certainly a decent and honest man.
***
`<Item id="2" name="Cancel"/>` in the [language files](https://github.com/pnedev/compare-plugin/files/1997930/TEST.zip) appears once within the diffs but multiple times in the entire files. 
Currently they are marked as Moved. But they are not a pair: 879 is under "Run a Macro Multiple Times" and 471 is under "ShortcutMapperSubDialg".
In this case the Moved mark is not helpful but rather confusing.

So there are three categories of potential Moved:

1) A line/block appears once in in each file and in the *entire file*. - Certain Moved: Marked as Moved.
2) A line/block appears multiple times within the diffs. - Question Moved: Not marked as Moved.

3) A line/block appears once in each file within the diffs but multiple times in the entire file(s). - ???

The reason you don't mark category `2` as Moved is that it wouldn't be helpful but rather confusing.
IOW: "Less is more.".
My suggestion was to apply "Less is more" to category `3` as well and not mark `<Item id="2" name="Cancel"/>` as Moved.

As you understand. I'm not going to argue. :)

Thanks again.
BR

**Comment by pnedev** [comment]
Hello Yaron,

The changed local mark you provided is in png format. Could you please send it like the other marks - as .h include file? Thank you, much appreciated.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

Oops, sorry.
[ChangedLocal.zip](https://github.com/pnedev/compare-plugin/files/2159765/ChangedLocal.zip)

Good to return to Move. :)
Thank you and best of luck.

**Comment by pnedev** [comment]
Hello Yaron,

Thank you, the icon is great :+1: 

I need to fix something else regarding move detection before I send you the binary to try.
I just didn't get the chance to do that this week.
Will write as soon as I can.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

I'm glad you like the icon. Thank you.

> I need to fix something else regarding move detection

Do you mean the Moved mark seen in https://github.com/pnedev/compare-plugin/issues/138#issuecomment-400301558? :)

> Will write as soon as I can.

Whenever you can. Your work is highly appreciated.
BR

**Comment by pnedev** [comment]
Hello Yaron,

Could you please try **[this](https://ci.appveyor.com/api/buildjobs/0ew1pi9lslwd9rmq/artifacts/ComparePlugin.dll)** binary whenever you find the time? It should include all marks and move detection changes we've discussed so far unless I have forgotten something.

Thank you, your help is highly appreciated.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

![default](https://user-images.githubusercontent.com/10229320/42847261-0de4c6d6-8a24-11e8-9b4f-40a3d6ac6a9e.png)

Any words would be an anticlimax, wouldn't they?
Well, let them be.

I've tested it quite thoroughly and both Moved and Local/Global-Added/Removed are implemented perfectly. A feast for the eyes.

Your hard work is indeed highly appreciated. Thank you very much.
***
Have you decided to mark a line as Moved if it appears once in the diffs but multiple times in the entire file(s)?
That's a legit approach and I wrote I wouldn't argue about that. :)
Still, allow me to share a thought which crossed my mind while testing the latest version.   

In the Local/Global-Added/Removed marks, your reference is the entire file including matches.
I think that to some degree it makes the other Moved approach more legit and perhaps even more correct.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

> Yes, indeed. My reasons are still the same...

:+1: 

I've come across an interesting case.
A single empty line is marked as [Moved](https://github.com/pnedev/compare-plugin/files/2205949/pair4.zip).
Shouldn't a sequence of two empty lines be marked as [Block-Moved](https://github.com/pnedev/compare-plugin/files/2205942/test_cmp.zip) (187-188 in `prefsFF.js` and 88-89 in `prefsFFbackup.js`)?

Thanks again for your excellent work.
BR

**Comment by pnedev** [comment]
Hello Yaron,

> A single empty line is marked as Moved.

Yes, it is expected as it is indeed moved.

> Shouldn't a sequence of two empty lines ...

Yes, definitely. A very good catch :+1: 
I'll investigate that, thank you.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

>  If a block is removed from one file and added to the other ONCE then it is definitely moved. In any other case it will be marked as locally changed.

Having read your reply again, I'm not sure we understand each other correctly.
My suggestion was to mark a line/block as Moved only if it's found ONCE in each file; *once in the entire file*.
So, for example, in `pair4` the empty line should not be marked as Moved: it appears more than once in the entire file. 
***
>Yes, it is expected as it is indeed moved.
> Yes, definitely. A very good catch.

This is also related to the previous point.
If you understood it correctly, how about marking empty lines as Moved only if they're part of a non-empty-lines block?

Thank you also for the NavBar fixes. :+1: 
You know I don't use it. :)

BR

**Comment by pnedev** [comment]
Hello Yaron,

I got confused here:

> My suggestion was to mark a line/block as Moved only if it's found ONCE in each file; once in the entire file.

But here is a part of a previous discussion:

> Have you decided to mark a line as Moved if it appears once in the diffs but multiple times in the entire file(s)?
That's a legit approach and I wrote I wouldn't argue about that.

This is how the move detection works right now (except the double-empty lines problem you caught).
I truly respect you opinion and it is highly appreciated but anyway I believe the move detection should focus ONLY on the diffs and not the entire file contents. If a line/block is added once in one file and removed once in the other then it is moved from one place to another despite the fact that it exists elsewhere (but matches between files). The other approach would mean that an added/removed block should be marked as such only if it exists once which is definitely not legit. That is uniqueness check, not compare.

I'll fix the other problem, thank you.

BR

**Comment by pnedev** [comment]
Hello Yaron,

> We've had quite a few (expected) misunderstanding during this long discussion...

It has been quite useful and fruitful actually. It just took ages because of my unavailability which brought more misunderstanding :)
Thank you.

I'll write when I fix the double-empty lines issue.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

> It has been quite useful and fruitful actually.

Indeed. The result is excellent.

> I'll write when I fix the double-empty lines issue.

Thank you.

BR

**Comment by pnedev** [comment]
Hello Yaron,

I haven't written for a long time, I hope you are doing well.

I have finally fixed the double-empty-lines issue and I think the move detection algorithm is behaving as expected now. Here is the x86 binary: https://ci.appveyor.com/api/buildjobs/tqcs7plx0t87aryf/artifacts/ComparePlugin.dll .

I'll wait for your confirmation first and then will close this ancient legendary thread.

Thank you and
BR

**Comment by Yaron10** [comment]
Hello Pavel,

I'm fine, thank you.
I hope you're doing well too.

Great work. :+1: 
Well, this long discussion has come to an end.
Thank you very much.

BR
